### PR TITLE
Fix port cutting z position

### PR DESCRIPTION
**Pull request by billingsley-john**
## Proposed changes

This PR partially replaces #610 by implementing the z_pos - center_point changes to the port cutters. This PR does NOT add automatic port cutting to reactors (like #610) - this will be done in a separate PR

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [ ] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Shimwell** [comment]
Smaller PRs can be easier to merge in so thanks for this change. I shall take a look soon

File: paramak/parametric_components/port_cutters_circular.py [hunk 1]
```diff
@@ -1,36 +1,37 @@
 
 from paramak import ExtrudeCircleShape
 
+from typing import Optional
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_circular.py:2]: Code inspector says this should be placed before paramak import :-)

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 2]
```diff
@@ -1,16 +1,18 @@
 
 from paramak import ExtrudeStraightShape
 
+from typing import Optional
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:2]: should be placed before paramak import

File: paramak/parametric_components/port_cutters_circular.py [hunk 3]
```diff
@@ -1,36 +1,37 @@
...
-        material_tag (str, optional): defaults to "circular_port_cutter_mat".
-        extrusion_start_offset (float, optional): the distance between 0 and
+        center_point: center point of the port cutter. Defaults to (0, 0).
+        radius: radius (cm) of port cutter.
+        distance: extruded distance (cm) of the port cutter.
+        stp_filename: defaults to "PortCutterCircular.stp".
+        stl_filename: defaults to "PortCutterCircular.stl".
+        name: defaults to "circular_port_cutter".
+        material_tag: defaults to "circular_port_cutter_mat".
+        extrusion_start_offset: the distance between 0 and
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_circular.py]: Parameters order in the docstrings and in the __init__ should be the same.

File: paramak/parametric_components/port_cutters_circular.py [hunk 4]
```diff
@@ -1,36 +1,37 @@
...
-        rotation_axis="Z",
-        extrusion_start_offset=1.,
-        stp_filename="PortCutterCircular.stp",
-        stl_filename="PortCutterCircular.stl",
-        name="circular_port_cutter",
-        material_tag="circular_port_cutter_mat",
+        radius: float,
+        distance: float,
+        center_point: Optional[tuple] = (0, 0),
+        workplane: Optional[str] = "ZY",
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_circular.py:32]: Non default value should appear in the docstrings

File: paramak/parametric_components/port_cutters_circular.py [hunk 5]
```diff
@@ -1,36 +1,37 @@
...
+        rotation_axis: Optional[str] = "Z",
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_circular.py:33]: Non default value should appear in the docstrings

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 6]
```diff
@@ -1,16 +1,18 @@
...
 
     Args:
-        z_pos (float): Z position (cm) of the port
-        height (float): height (cm) of the port
-        width (float): width (cm) of the port
-        distance (float): extruded distance (cm) of the cutter
+        center_point: Center point of the port cutter. Defaults to (0, 0).
+        height: height (cm) of the port cutter.
+        width: width (cm) of the port cutter.
+        distance: extruded distance (cm) of the port cutter.
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py]: Parameters order in the docstrings and in the init should be the same.

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 7]
```diff
@@ -24,18 +26,18 @@ class PortCutterRectangular(ExtrudeStraightShape):
...
-        fillet_radius=None,
-        stp_filename="PortCutterRectangular.stp",
-        stl_filename="PortCutterRectangular.stl",
-        name="rectangular_port_cutter",
-        material_tag="rectangular_port_cutter_mat",
+        height: float,
+        width: float,
+        distance: float,
+        center_point: Optional[tuple] = (0, 0),
+        workplane: Optional[str] = "ZY",
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:37]: Non default value should appear in the docstrings

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 8]
```diff
@@ -24,18 +26,18 @@ class PortCutterRectangular(ExtrudeStraightShape):
...
+        rotation_axis: Optional[str] = "Z",
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:38]: Non default value should appear in the docstrings

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 9]
```diff
@@ -52,11 +54,11 @@ def __init__(
             **kwargs
         )
 
-        self.z_pos = z_pos
+        self.center_point = center_point
         self.height = height
         self.width = width
         self.fillet_radius = fillet_radius
-        self.add_fillet()
+        # self.add_fillet()
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py]: should this be removed ?

File: tests/test_parametric_components/test_PortCutterCircular.py [hunk 10]
```diff
@@ -1,19 +1,56 @@
-
+import math
 import unittest
 
 import paramak
 
+import pytest
+
+
+class test_component(unittest.TestCase):
```
Comment (RemDelaporteMathurin) [inline tests/test_parametric_components/test_PortCutterCircular.py]: Code inspector wants it to be PascalCase style
```suggestion
class TestComponent(unittest.TestCase):
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Some really tiny comments :-)

**Comment by billingsley-john** [comment]
Thanks @RemDelaporteMathurin for the suggested changes

File: paramak/parametric_components/port_cutters_circular.py (hunk 4 above)
Comment (billingsley-john) [reply paramak/parametric_components/port_cutters_circular.py:32]: Radius and distance are documented in the docstrings. Apologies if I have misunderstood!

Comment (RemDelaporteMathurin) [reply paramak/parametric_components/port_cutters_circular.py:32]: Sorry. `center_point` and `workplane` are not documented in the docstrings. Similarly to arguments like name, material_tag, etc. their default value should appear
```
center_point: Defaults to (0, 0).
workplane: Defaults to "ZY".
```

Comment (billingsley-john) [reply paramak/parametric_components/port_cutters_circular.py:32]: Ah sorry, I understand. Yes, you're right, they should be added.

**Comment by billingsley-john** [comment]
Think this is ready to merge unless anyone has any suggestions? @RemDelaporteMathurin @Shimwell

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 11]
```diff
@@ -52,22 +58,39 @@ def __init__(
...
         self.height = height
         self.width = width
+        self.center_point = center_point
         self.fillet_radius = fillet_radius
-        self.add_fillet()
 
     def find_points(self):
+        if self.workplane[0] < self.workplane[1]:
+            parameter_1 = self.width
+            parameter_2 = self.height
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:69]: These lines aren't tested. Is there a way of adapting some of the tests?

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 12]
```diff
@@ -52,22 +58,39 @@ def __init__(
...
+                   self.center_point[1]) for e in points]
+
         self.points = points
 
-    def add_fillet(self):
+    def add_fillet(self, solid):
+        if "X" not in self.workplane:
+            filleting_edge = "|X"
+        if "Y" not in self.workplane:
+            filleting_edge = "|Y"
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:89]: Not tested

File: paramak/parametric_components/port_cutters_rectangular.py [hunk 13]
```diff
@@ -52,22 +58,39 @@ def __init__(
...
+        if "Z" not in self.workplane:
+            filleting_edge = "|Z"
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/port_cutters_rectangular.py:91]: Not tested

File: paramak/parametric_components/port_cutters_rectangular.py (hunk 11 above)
Comment (billingsley-john) [reply paramak/parametric_components/port_cutters_rectangular.py:69]: Thanks for pointing this out. Tests will be added

**Comment by billingsley-john** [comment]
Think this should be ready for merging @Shimwell @RemDelaporteMathurin once tests pass

**Review by RemDelaporteMathurin — APPROVED** [review]

**Pull request merged**